### PR TITLE
Add nvFuser support for torch.Tensor.view

### PR DESCRIPTION
**Pull request by IvanYashchuk**
This is an alternative to https://github.com/pytorch/pytorch/pull/83739. While PrimTorch has `view` as a reference, we would like to use nvFuser's implementation for `view` for now. Later we might transition to PrimTorch's `torch._refs.view`.

See `test_nvprims_view` for examples of things that are now sent to nvFuser. Note that nvFuser's `view` is a copy-like operation.

### CODE REVIEW

File: test/test_prims.py [hunk 1]
```diff
@@ -481,6 +481,50 @@ def _wrapper(a):
         )
         self.assertTrue(includes_nvprims_var_mean)
 
+    @onlyCUDA
+    @skipCUDAIfRocm
+    @dtypes(torch.float16, torch.float32)
+    def test_nvprims_view(self, device, dtype):
```
Comment (IvanYashchuk) [inline test/test_prims.py:487]: - [ ] Add `aten._unsafe_view`

File: torch/_prims/context.py [hunk 2]
```diff
@@ -230,6 +230,21 @@ def _is_var_mean(self, func):
...
+            "torch.Tensor.view" == torch.overrides.resolve_name(func)
+            or "torch.Tensor.reshape" == torch.overrides.resolve_name(func)
+            or "torch.view_copy" == torch.overrides.resolve_name(func)
+            or "torch.reshape" == torch.overrides.resolve_name(func)
+            or (
+                (
+                    isinstance(func, torch._ops.OpOverload)
+                    or isinstance(func, torch._ops.OpOverloadPacket)
+                )
+                and "aten.view" in str(func)
```
Comment (jjsjann123) [inline torch/_prims/context.py:244]: The last entry is really general. I think we'll also match `torch.ops.aten.view.dtype` here, which takes a slightly different signature than what ` torch.ops.nvprims.view` supports.

A quick regex search with `func: view` in `native_functions.yaml` shows a bunch other ops in the same situation. Is there some other pieces in the stack that protects us from the overloads?

Comment (IvanYashchuk) [reply torch/_prims/context.py:244]: I will change this to be specific.

File: torch/_prims/nvfuser_prims.py [hunk 3]
```diff
@@ -398,9 +408,43 @@ def _var_mean_autograd(
...
+
+    nvprim_impl.impl(name, _prim_impl)
+
+    prim_packet = torch.ops.nvprims.view
+    prim = prim_packet.default
+
+    def _view_no_original_shape_overload_impl(a, shape):
+        if list(a.shape) == list(shape):
+            return torch.ops.nvprims.view_of(a)
+        return prim(a, a.shape, shape)
```
Comment (IvanYashchuk) [inline torch/_prims/nvfuser_prims.py:433]:
```suggestion
        return torch.ops.nvprims.view.default(a, a.shape, shape)
```

**Comment by mruberry** [comment]
The OOMs seem like they could be relevant?

**Comment by IvanYashchuk** [comment]
I hope not, I will check that.

File: torch/csrc/jit/codegen/cuda/python_frontend/fusion_record.h [hunk 4]
```diff
@@ -86,6 +86,31 @@ struct OpRecord : RecordFunctor {
...
+      std::vector<int64_t>& new_sizes)
+      : RecordFunctor(std::move(_args), std::move(_outputs)),
+        original_sizes_(std::move(original_sizes)),
+        new_sizes_(std::move(new_sizes)) {}
+  virtual ~ViewOpRecord() = default;
+
+  void operator()(FusionDefinition& fd) final {
+    auto arg = fd.getFusionState(args.at(0))->template as<TensorView>();
+    auto output =
+        torch::jit::fuser::cuda::view(arg, original_sizes_, new_sizes_);
```
[the remark this answers is not in the document]
Comment (kevinstephano) [reply torch/csrc/jit/codegen/cuda/python_frontend/fusion_record.h:103]: What would you like to guard?  I think an issue is you aren't really looking at the Record used for caching.  The new Record's equality operator is probably what you would consider the guard and that is not in upstream, yet.

**Comment by IvanYashchuk** [comment]
Recent log with OOM:
```
2022-09-29T14:59:28.8636208Z test_ops.py::TestCommonCUDA::test_python_ref_executor_ops_nvprims_view_executor_nvfuser_cuda_float32 PASSED [ 51%]
2022-09-29T14:59:28.8636474Z test_ops.py::TestCommonCUDA::test_python_ref_executor_ops_nvprims_view_executor_nvfuser_cuda_int64 SKIPPED (skipped for speed) [ 51%]
2022-09-29T14:59:28.8636938Z test_ops.py::TestCommonCUDA::test_python_ref_executor_ops_nvprims_view_executor_nvfuser_cuda_uint8 SKIPPED (nvfuser doesn't support dtype torch.uint8) [ 51%]
2022-09-29T14:59:28.8637159Z test_ops.py::TestCommonCUDA::test_python_ref_meta__refs_abs_cuda_complex128 PASSED [ 51%]
2022-09-29T14:59:28.8637414Z test_ops.py::TestCommonCUDA::test_python_ref_meta__refs_abs_cuda_float16 PASSED [ 51%]
2022-09-29T14:59:28.8637628Z test_ops.py::TestCommonCUDA::test_python_ref_meta__refs_abs_cuda_float64 PASSED [ 51%]
2022-09-29T14:59:28.8637838Z test_ops.py::TestCommonCUDA::test_python_ref_meta__refs_abs_cuda_int32 PASSED [ 51%]
2022-09-29T14:59:28.8638042Z test_ops.py::TestCommonCUDA::test_python_ref_meta__refs_abs_cuda_int8 PASSED [ 51%]
2022-09-29T14:59:28.8638346Z test_ops.py::TestCommonCUDA::test_python_ref_meta__refs_acos_cuda_bfloat16 PASSED [ 51%]
2022-09-29T14:59:28.8638574Z test_ops.py::TestCommonCUDA::test_python_ref_meta__refs_acos_cuda_complex128 PASSED [ 51%]
2022-09-29T14:59:28.8638781Z test_ops.py::TestCommonCUDA::test_python_ref_meta__refs_acos_cuda_complex32 PASSED [ 51%]
2022-09-29T14:59:28.8638990Z test_ops.py::TestCommonCUDA::test_python_ref_meta__refs_acos_cuda_float16 PASSED [ 51%]
2022-09-29T14:59:28.8639193Z test_ops.py::TestCommonCUDA::test_python_ref_meta__refs_acos_cuda_int16 PASSED [ 51%]
2022-09-29T14:59:28.8639402Z test_ops.py::TestCommonCUDA::test_python_ref_meta__refs_acosh_cuda_bfloat16 PASSED [ 51%]
2022-09-29T14:59:28.8639609Z test_ops.py::TestCommonCUDA::test_python_ref_meta__refs_acosh_cuda_complex64 RERUN [ 51%]
2022-09-29T14:59:28.8639814Z test_ops.py::TestCommonCUDA::test_python_ref_meta__refs_acosh_cuda_complex64 RERUN [ 51%]
2022-09-29T14:59:28.8640028Z test_ops.py::TestCommonCUDA::test_python_ref_meta__refs_acosh_cuda_complex64 FAILED [ 51%]
2022-09-29T14:59:28.8640052Z 
2022-09-29T14:59:28.8640207Z =================================== FAILURES ===================================
2022-09-29T14:59:28.8640394Z ________ TestCommonCUDA.test_python_ref_meta__refs_acosh_cuda_complex64 ________
2022-09-29T14:59:28.8640528Z Traceback (most recent call last):
2022-09-29T14:59:28.8640735Z   File "/var/lib/jenkins/workspace/test/test_ops.py", line 185, in test_python_ref_meta
2022-09-29T14:59:28.8640934Z     for sample in op.reference_inputs(device, dtype, requires_grad=False):
2022-09-29T14:59:28.8641297Z   File "/opt/conda/lib/python3.10/site-packages/torch/autograd/grad_mode.py", line 64, in generator_context
2022-09-29T14:59:28.8641425Z     response = gen.send(request)
2022-09-29T14:59:28.8641828Z   File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/opinfo/core.py", line 2254, in reference_inputs_elementwise_unary
2022-09-29T14:59:28.8641964Z     yield sample.noncontiguous()
2022-09-29T14:59:28.8642308Z   File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/opinfo/core.py", line 262, in noncontiguous
2022-09-29T14:59:28.8642456Z     return self.transform(to_noncontiguous)
2022-09-29T14:59:28.8642802Z   File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/opinfo/core.py", line 220, in transform
2022-09-29T14:59:28.8642970Z     tt(self.input),
2022-09-29T14:59:28.8643304Z   File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/opinfo/core.py", line 207, in tt
2022-09-29T14:59:28.8643406Z     return _tt(t)
2022-09-29T14:59:28.8643735Z   File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/opinfo/core.py", line 204, in _tt
2022-09-29T14:59:28.8643836Z     return f(t)
2022-09-29T14:59:28.8644178Z   File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/opinfo/core.py", line 256, in to_noncontiguous
2022-09-29T14:59:28.8644303Z     return noncontiguous_like(t)
2022-09-29T14:59:28.8644668Z   File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 2942, in noncontiguous_like
2022-09-29T14:59:28.8644804Z     result = t.new_empty(t.shape + (2,))
2022-09-29T14:59:28.8645374Z torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 16.00 MiB (GPU 0; 7.44 GiB total capacity; 255.20 MiB already allocated; 11.19 MiB free; 1.64 GiB allowed; 682.00 MiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
2022-09-29T14:59:28.8645771Z - generated xml file: /var/lib/jenkins/workspace/test/test-reports/python-pytest/test_ops/test_ops-aeb0f2696ff072c7.xml -
2022-09-29T14:59:28.8645934Z =========================== short test summary info ============================
2022-09-29T14:59:28.8646151Z FAILED test_ops.py::TestCommonCUDA::test_python_ref_meta__refs_acosh_cuda_complex64
2022-09-29T14:59:28.8646363Z !!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!
2022-09-29T14:59:28.8646595Z = 1 failed, 3175 passed, 1163 skipped, 14 deselected, 67 xfailed, 29 warnings, 2 rerun in 1226.94s (0:20:26) =
```

File: torch/csrc/jit/codegen/cuda/python_frontend/fusion_record.h [hunk 5]
```diff
@@ -277,6 +280,99 @@ struct OpRecord : RecordFunctor {
...
+struct ViewOpRecord : RecordFunctor {
+  ViewOpRecord(
+      std::vector<State> _args,
+      std::vector<State> _outputs,
+      std::vector<int64_t>& original_shape,
+      std::vector<int64_t>& new_shape)
+      : RecordFunctor(
+            std::move(_args),
+            std::move(_outputs),
+            "view",
```
Comment (kevinstephano) [inline torch/csrc/jit/codegen/cuda/python_frontend/fusion_record.h]: Could you change the name from `view` to `ops.view`?

**Review by kevinstephano — APPROVED** [review]

**Review by mruberry — APPROVED** [review]
Stamped!

File: torch/_prims/context.py [hunk 6]
```diff
@@ -301,12 +313,15 @@ def __torch_function__(
         if self._is_var_mean(orig_func):
             return torch.ops.nvprims.var_mean(*args, **kwargs)
 
+        if self._is_view_or_reshape(orig_func):
+            return torch.ops.nvprims.view(*args, **kwargs)
```
Comment (IvanYashchuk) [inline torch/_prims/context.py:317]: We need this to support `tensor.view(1, 2, 3, 4, 5, 6)`
```suggestion
        if self._is_view_or_reshape(orig_func):
            a, *shape = args
            shape = torch._prims_common.extract_shape_from_varargs(shape, validate=False)
            if len(kwargs) > 0:
                warn("view has ignored kwargs!")
            return torch.ops.nvprims.view(a, shape)
```

**Pull request closed without merging**